File: ios/Classes/models/map/here_map/LanguageCodeConverter.swift
/*
 * Copyright (C) 2019-2024 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

import Foundation
import heresdk

// Converts from HERE SDK's LanguageCode to Locale and vice versa.
// Both language and country must be set, if available.
class LanguageCodeConverter {

    // Language is always set, region may not be set.
    private static let languageCodeMap = [

        /// English (United States)
        LanguageCode.enUs: Locale(identifier: "en-US"),

        /// Afrikaans
        LanguageCode.afZa: Locale(identifier: "af-ZA"),

        /// Albanian
        LanguageCode.sqAl: Locale(identifier: "sq-AL"),

        /// Amharic (Ethiopia)
        LanguageCode.amEt: Locale(identifier: "am-ET"),

        /// Arabic (Saudi Arabia)
        LanguageCode.arSa: Locale(identifier: "ar-SA"),

        /// Armenian
        LanguageCode.hyAm: Locale(identifier: "hy-AM"),

        /// Assamese (India)
        LanguageCode.asIn: Locale(identifier: "as-IN"),

        /// Azeri - Latin
        LanguageCode.azLatnAz: Locale(identifier: "az-LATN_AZ"),

        /// Bangla (Bangladesh)
        LanguageCode.bnBd: Locale(identifier: "bn-BD"),

        /// Bangla (India)
        LanguageCode.bnIn: Locale(identifier: "bn-IN"),

        /// Basque
        LanguageCode.euEs: Locale(identifier: "eu-ES"),

        /// Belarusian
        LanguageCode.beBy: Locale(identifier: "be-BY"),

        /// Bosnian - Latin
        LanguageCode.bsLatnBa: Locale(identifier: "bs-LATN_BA"),

        /// Bulgarian
        LanguageCode.bgBg: Locale(identifier: "bg-BG"),

        /// Catalan (Spain)
        LanguageCode.caEs: Locale(identifier: "ca-ES"),

        /// Central Kurdish - Arabic
        LanguageCode.kuArab: Locale(identifier: "ku-ARAB"),

        /// Chinese (Simplified China)
        LanguageCode.zhCn: Locale(identifier: "zh-CN"),

        /// Chinese (Traditional Hong Kong)
        LanguageCode.zhHk: Locale(identifier: "zh-HK"),

        /// Chinese (Traditional Taiwan)
        LanguageCode.zhTw: Locale(identifier: "zh-TW"),

        /// Croatian
        LanguageCode.hrHr: Locale(identifier: "hr-HR"),

        /// Czech
        LanguageCode.csCz: Locale(identifier: "cs-CZ"),

        /// Danish
        LanguageCode.daDk: Locale(identifier: "da-DK"),

        /// Dari - Arabic (Afghanistan)
        LanguageCode.prsArabAf: Locale(identifier: "prs-ARAB_AF"),

        /// Dutch
        LanguageCode.nlNl: Locale(identifier: "nl-NL"),

        /// English (British)
        LanguageCode.enGb: Locale(identifier: "en-GB"),

        /// Estonian
        LanguageCode.etEe: Locale(identifier: "et-EE"),

        /// Farsi (Iran)
        LanguageCode.faIr: Locale(identifier: "fa-IR"),

        /// Filipino
        LanguageCode.filPh: Locale(identifier: "fil-PH"),

        /// Finnish
        LanguageCode.fiFi: Locale(identifier: "fi-FI"),

        /// French
        LanguageCode.frFr: Locale(identifier: "fr-FR"),

        /// French (Canada)
        LanguageCode.frCa: Locale(identifier: "fr-CA"),

        /// Galician
        LanguageCode.glEs: Locale(identifier: "gl-ES"),

        /// Georgian
        LanguageCode.kaGe: Locale(identifier: "ka-GE"),

        /// German
        LanguageCode.deDe: Locale(identifier: "de-DE"),

        /// Greek
        LanguageCode.elGr: Locale(identifier: "el-GR"),

        /// Gujarati (India)
        LanguageCode.guIn: Locale(identifier: "gu-IN"),

        /// Hausa - Latin (Nigeria)
        LanguageCode.haLatnNg: Locale(identifier: "ha-LATN_NG"),

        /// Hebrew
        LanguageCode.heIl: Locale(identifier: "he-IL"),

        /// Hindi
        LanguageCode.hiIn: Locale(identifier: "hi-IN"),

        /// Hungarian
        LanguageCode.huHu: Locale(identifier: "hu-HU"),

        /// Icelandic
        LanguageCode.isIs: Locale(identifier: "is-IS"),

        /// Igbo - Latin (Nigera)
        LanguageCode.igLatnNg: Locale(identifier: "ig-LATN_NG"),

        /// Indonesian (Bahasa)
        LanguageCode.idId: Locale(identifier: "id-ID"),

        /// Irish
        LanguageCode.gaIe: Locale(identifier: "ga-IE"),

        /// IsiXhosa
        LanguageCode.xh: Locale(identifier: "xh"),

        /// IsiZulu (South Africa)
        LanguageCode.zuZa: Locale(identifier: "zu-ZA"),

        /// Italian
        LanguageCode.itIt: Locale(identifier: "it-IT"),

        /// Japanese
        LanguageCode.jaJp: Locale(identifier: "ja-JP"),

        /// Kannada (India)
        LanguageCode.knIn: Locale(identifier: "kn-IN"),

        /// Kazakh
        LanguageCode.kkKz: Locale(identifier: "kk-KZ"),

        /// Khmer (Cambodia)
        LanguageCode.kmKh: Locale(identifier: "km-KH"),

        /// K'iche' - Latin (Guatemala)
        LanguageCode.qucLatnGt: Locale(identifier: "quc-LATN_GT"),

        /// Kinyarwanda (Rwanda)
        LanguageCode.rwRw: Locale(identifier: "rw-RW"),

        /// KiSwahili
        LanguageCode.sw: Locale(identifier: "sw"),

        /// Konkani (India)
        LanguageCode.kokIn: Locale(identifier: "kok-IN"),

        /// Korean
        LanguageCode.koKr: Locale(identifier: "ko-KR"),

        /// Kyrgyz - Cyrillic
        LanguageCode.kyCyrlKg: Locale(identifier: "ky-CYRL_KG"),

        /// Latvian
        LanguageCode.lvLv: Locale(identifier: "lv-LV"),

        /// Lithuanian
        LanguageCode.ltLt: Locale(identifier: "lt-LT"),

        /// Luxembourgish
        LanguageCode.lbLu: Locale(identifier: "lb-LU"),

        /// Macedonian
        LanguageCode.mkMk: Locale(identifier: "mk-MK"),

        /// Malay (Bahasa)
        LanguageCode.msMy: Locale(identifier: "ms-MY"),

        /// Malayalam (India)
        LanguageCode.mlIn: Locale(identifier: "ml-IN"),

        /// Maltese  (Malta)
        LanguageCode.mtMt: Locale(identifier: "mt-MT"),

        /// Maori - Latin (New Zealand)
        LanguageCode.miLatnNz: Locale(identifier: "mi-LATN_NZ"),

        /// Marathi (India)
        LanguageCode.mrIn: Locale(identifier: "mr-IN"),

        /// Mongolian - Cyrillic
        LanguageCode.mnCyrlMn: Locale(identifier: "mn-CYRL_MN"),

        /// Nepali (Nepal)
        LanguageCode.neNp: Locale(identifier: "ne-NP"),

        /// Norwegian (Bokmål)
        LanguageCode.nbNo: Locale(identifier: "nb-NO"),

        /// Norwegian (Nynorsk)
        LanguageCode.nnNo: Locale(identifier: "nn-NO"),

        /// Odia (India)
        LanguageCode.orIn: Locale(identifier: "or-IN"),

        /// Polish
        LanguageCode.plPl: Locale(identifier: "pl-PL"),

        /// Portuguese (Brazil)
        LanguageCode.ptBr: Locale(identifier: "pt-BR"),

        /// Portuguese (Portugal)
        LanguageCode.ptPt: Locale(identifier: "pt-PT"),

        /// Punjabi - Gurmukhi
        LanguageCode.paGuru: Locale(identifier: "pa-GURU"),

        /// Punjabi - Arabic
        LanguageCode.paArab: Locale(identifier: "pa-ARAB"),

        /// Quechua - Latin (Peru)
        LanguageCode.quLatnPe: Locale(identifier: "qu-LATN_PE"),

        /// Romanian
        LanguageCode.roRo: Locale(identifier: "ro-RO"),

        /// Russian
        LanguageCode.ruRu: Locale(identifier: "ru-RU"),

        /// Scottish Gaelic - Latin
        LanguageCode.gdLatnGb: Locale(identifier: "gd-LATN_GB"),

        /// Serbian - Cyrillic (Bosnia)
        LanguageCode.srCyrlBa: Locale(identifier: "sr-CYRL_BA"),

        /// Serbian - Cyrillic (Serbia)
        LanguageCode.srCyrlRs: Locale(identifier: "sr-CYRL_RS"),

        /// Serbian - Latin (Serbia)
        LanguageCode.srLatnRs: Locale(identifier: "sr-LATN_RS"),

        /// Sesotho Sa Leboa (South Africa)
        LanguageCode.nsoZa: Locale(identifier: "nso-ZA"),

        /// Setswana
        LanguageCode.tn: Locale(identifier: "tn"),

        /// Sindhi - Arabic
        LanguageCode.sdArab: Locale(identifier: "sd-ARAB"),

        /// Sinhala (Sri Lanka)
        LanguageCode.siLk: Locale(identifier: "si-LK"),

        /// Slovak
        LanguageCode.skSk: Locale(identifier: "sk-SK"),

        /// Slovenian
        LanguageCode.slSi: Locale(identifier: "sl-SI"),

        /// Spanish (Mexico)
        LanguageCode.esMx: Locale(identifier: "es-MX"),

        /// Spanish (Spain)
        LanguageCode.esEs: Locale(identifier: "es-ES"),

        /// Swedish
        LanguageCode.svSe: Locale(identifier: "sv-SE"),

        /// Tajik - Cyrillic
        LanguageCode.tgCyrlTj: Locale(identifier: "tg-CYRL_TJ"),

        /// Tamil
        LanguageCode.ta: Locale(identifier: "ta"),

        /// Tatar - Cyrillic (Russia)
        LanguageCode.ttCyrlRu: Locale(identifier: "tt-CYRL_RU"),

        /// Telugu (India)
        LanguageCode.teIn: Locale(identifier: "te-IN"),

        /// Thai
        LanguageCode.thTh: Locale(identifier: "th-TH"),

        /// Tigrinya (Ethiopia)
        LanguageCode.tiEt: Locale(identifier: "ti-ET"),

        /// Turkish
        LanguageCode.trTr: Locale(identifier: "tr-TR"),

        /// Turkmen - Latin
        LanguageCode.tkLatnTm: Locale(identifier: "tk-LATN_TM"),

        /// Ukrainian
        LanguageCode.ukUa: Locale(identifier: "uk-UA"),

        /// Urdu
        LanguageCode.ur: Locale(identifier: "ur"),

        /// Uyghur - Arabic
        LanguageCode.ugArab: Locale(identifier: "ug-ARAB"),

        /// Uzbek - Cyrillic
        LanguageCode.uzCyrlUz: Locale(identifier: "uz-CYRL_UZ"),

        /// Uzbek - Latin
        LanguageCode.uzLatnUz: Locale(identifier: "uz-LATN_UZ"),

        /// Valencian (Spain)
        LanguageCode.catEs: Locale(identifier: "cat-ES"),

        /// Vietnamese
        LanguageCode.viVn: Locale(identifier: "vi-VN"),

        /// Welsh
        LanguageCode.cyGb: Locale(identifier: "cy-GB"),

        /// Wolof - Latin
        LanguageCode.woLatn: Locale(identifier: "wo-LATN"),

        /// Yoruba - Latin
        LanguageCode.yoLatn: Locale(identifier: "yo-LATN")]

    public static func getLocale(languageCode: LanguageCode) -> Locale {
        guard let index = languageCodeMap.index(forKey: languageCode) else {
            // Should never happen, unless the languageCodeMap was not updated
            // to support the latest LanguageCodes from HERE SDK.
            print("LanguageCode not found. Falling Back to en-US.")
            return Locale(identifier: "en-US")
        }

        return languageCodeMap[index].value
    }

    public static func getLanguageCode(locale: Locale) -> LanguageCode {
        let language = locale.languageCode
        let country = locale.regionCode

        for (languageCodeEntry, localeEntry) in languageCodeMap {
            if country == nil {
                if language == localeEntry.languageCode {
                    return languageCodeEntry
                }
            } else {
                if language == localeEntry.languageCode && country == localeEntry.regionCode {
                    return languageCodeEntry
                }
            }
        }

        print("LanguageCode not found. Falling back to enUs.")
        return LanguageCode.enUs
    }

    public static func getLanguageCode(identifier: String) -> LanguageCode {
        for (languageCodeEntry, localeEntry) in languageCodeMap {
            if identifier == localeEntry.identifier {
                return languageCodeEntry
            }
        }

        print("LanguageCode not found. Falling back to enUs.")
        return LanguageCode.enUs
    }
}

File: android/src/main/kotlin/com/oguzhnatly/flutter_carplay/models/map/here_map/LanguageCodeConverter.kt
/*
 * Copyright (C) 2019-2024 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */
package com.oguzhnatly.flutter_carplay.models.map.here_map

import android.util.Log
import com.here.sdk.core.LanguageCode
import java.util.Locale

// Converts from com.here.sdk.core.LanguageCode to java.util.Locale and vice versa.
// Both language and country must be set, if available.
object LanguageCodeConverter {
    private val TAG: String = LanguageCodeConverter::class.java.name

    private var languageCodeMap: HashMap<LanguageCode, Locale>? = null

    fun getLocale(languageCode: LanguageCode): Locale? {
        if (languageCodeMap == null) {
            initLanguageCodeMap()
        }

        if (languageCodeMap!!.containsKey(languageCode)) {
            return languageCodeMap!![languageCode]
        }

        // Should never happen, unless the languageCodeMap was not updated
        // to support the latest LanguageCodes from HERE SDK.
        Log.e(TAG, "LanguageCode not found. Falling Back to en-US.")
        return Locale("en", "US")
    }

    fun getLanguageCode(locale: Locale): LanguageCode {
        if (languageCodeMap == null) {
            initLanguageCodeMap()
        }

        val language = locale.language
        val country = locale.country

        for (languageCodeLocaleEntry in languageCodeMap!!.entries) {
            val localeEntry = (languageCodeLocaleEntry as Map.Entry<*, *>).value as Locale

            val languageEntry = localeEntry.language
            val countryEntry = localeEntry.country

            if (country == null) {
                if (language == languageEntry) {
                    return (languageCodeLocaleEntry as Map.Entry<*, *>).key as LanguageCode
                }
            } else {
                if (language == languageEntry && country == countryEntry) {
                    return (languageCodeLocaleEntry as Map.Entry<*, *>).key as LanguageCode
                }
            }
        }

        Log.e(TAG, "LanguageCode not found. Falling back to EN_US.")
        return LanguageCode.EN_US
    }

    // / Language is always set, country may not be set.
    private fun initLanguageCodeMap() {
        languageCodeMap = HashMap()

        /// English (United States)
        languageCodeMap!![LanguageCode.EN_US] = Locale("en", "US")

        /// Afrikaans
        languageCodeMap!![LanguageCode.AF_ZA] = Locale("af", "ZA")

        /// Albanian
        languageCodeMap!![LanguageCode.SQ_AL] = Locale("sq", "AL")

        /// Amharic (Ethiopia)
        languageCodeMap!![LanguageCode.AM_ET] = Locale("am", "ET")

        /// Arabic (Saudi Arabia)
        languageCodeMap!![LanguageCode.AR_SA] = Locale("ar", "SA")

        /// Armenian
        languageCodeMap!![LanguageCode.HY_AM] = Locale("hy", "AM")

        /// Assamese (India)
        languageCodeMap!![LanguageCode.AS_IN] = Locale("as", "IN")

        /// Azeri - Latin
        languageCodeMap!![LanguageCode.AZ_LATN_AZ] = Locale("az", "LATN_AZ")

        /// Bangla (Bangladesh)
        languageCodeMap!![LanguageCode.BN_BD] = Locale("bn", "BD")

        /// Bangla (India)
        languageCodeMap!![LanguageCode.BN_IN] = Locale("bn", "IN")

        /// Basque
        languageCodeMap!![LanguageCode.EU_ES] = Locale("eu", "ES")

        /// Belarusian
        languageCodeMap!![LanguageCode.BE_BY] = Locale("be", "BY")

        /// Bosnian - Latin
        languageCodeMap!![LanguageCode.BS_LATN_BA] = Locale("bs", "LATN_BA")

        /// Bulgarian
        languageCodeMap!![LanguageCode.BG_BG] = Locale("bg", "BG")

        /// Catalan (Spain)
        languageCodeMap!![LanguageCode.CA_ES] = Locale("ca", "ES")

        /// Central Kurdish - Arabic
        languageCodeMap!![LanguageCode.KU_ARAB] = Locale("ku", "ARAB")

        /// Chinese (Simplified China)
        languageCodeMap!![LanguageCode.ZH_CN] = Locale("zh", "CN")

        /// Chinese (Traditional Hong Kong)
        languageCodeMap!![LanguageCode.ZH_HK] = Locale("zh", "HK")

        /// Chinese (Traditional Taiwan)
        languageCodeMap!![LanguageCode.ZH_TW] = Locale("zh", "TW")

        /// Croatian
        languageCodeMap!![LanguageCode.HR_HR] = Locale("hr", "HR")

        /// Czech
        languageCodeMap!![LanguageCode.CS_CZ] = Locale("cs", "CZ")

        /// Danish
        languageCodeMap!![LanguageCode.DA_DK] = Locale("da", "DK")

        /// Dari - Arabic (Afghanistan)
        languageCodeMap!![LanguageCode.PRS_ARAB_AF] = Locale("prs", "ARAB_AF")

        /// Dutch
        languageCodeMap!![LanguageCode.NL_NL] = Locale("nl", "NL")

        /// English (British)
        languageCodeMap!![LanguageCode.EN_GB] = Locale("en", "GB")

        /// Estonian
        languageCodeMap!![LanguageCode.ET_EE] = Locale("et", "EE")

        /// Farsi (Iran)
        languageCodeMap!![LanguageCode.FA_IR] = Locale("fa", "IR")

        /// Filipino
        languageCodeMap!![LanguageCode.FIL_PH] = Locale("fil", "PH")

        /// Finnish
        languageCodeMap!![LanguageCode.FI_FI] = Locale("fi", "FI")

        /// French
        languageCodeMap!![LanguageCode.FR_FR] = Locale("fr", "FR")

        /// French (Canada)
        languageCodeMap!![LanguageCode.FR_CA] = Locale("fr", "CA")

        /// Galician
        languageCodeMap!![LanguageCode.GL_ES] = Locale("gl", "ES")

        /// Georgian
        languageCodeMap!![LanguageCode.KA_GE] = Locale("ka", "GE")

        /// German
        languageCodeMap!![LanguageCode.DE_DE] = Locale("de", "DE")

        /// Greek
        languageCodeMap!![LanguageCode.EL_GR] = Locale("el", "GR")

        /// Gujarati (India)
        languageCodeMap!![LanguageCode.GU_IN] = Locale("gu", "IN")

        /// Hausa - Latin (Nigeria)
        languageCodeMap!![LanguageCode.HA_LATN_NG] = Locale("ha", "LATN_NG")

        /// Hebrew
        languageCodeMap!![LanguageCode.HE_IL] = Locale("he", "IL")

        /// Hindi
        languageCodeMap!![LanguageCode.HI_IN] = Locale("hi", "IN")

        /// Hungarian
        languageCodeMap!![LanguageCode.HU_HU] = Locale("hu", "HU")

        /// Icelandic
        languageCodeMap!![LanguageCode.IS_IS] = Locale("is", "IS")

        /// Igbo - Latin (Nigera)
        languageCodeMap!![LanguageCode.IG_LATN_NG] = Locale("ig", "LATN_NG")

        /// Indonesian (Bahasa)
        languageCodeMap!![LanguageCode.ID_ID] = Locale("id", "ID")

        /// Irish
        languageCodeMap!![LanguageCode.GA_IE] = Locale("ga", "IE")

        /// IsiXhosa
        languageCodeMap!![LanguageCode.XH] = Locale("xh")

        /// IsiZulu (South Africa)
        languageCodeMap!![LanguageCode.ZU_ZA] = Locale("zu", "ZA")

        /// Italian
        languageCodeMap!![LanguageCode.IT_IT] = Locale("it", "IT")

        /// Japanese
        languageCodeMap!![LanguageCode.JA_JP] = Locale("ja", "JP")

        /// Kannada (India)
        languageCodeMap!![LanguageCode.KN_IN] = Locale("kn", "IN")

        /// Kazakh
        languageCodeMap!![LanguageCode.KK_KZ] = Locale("kk", "KZ")

        /// Khmer (Cambodia)
        languageCodeMap!![LanguageCode.KM_KH] = Locale("km", "KH")

        /// K'iche' - Latin (Guatemala)
        languageCodeMap!![LanguageCode.QUC_LATN_GT] = Locale("quc", "LATN_GT")

        /// Kinyarwanda (Rwanda)
        languageCodeMap!![LanguageCode.RW_RW] = Locale("rw", "RW")

        /// KiSwahili
        languageCodeMap!![LanguageCode.SW] = Locale("sw")

        /// Konkani (India)
        languageCodeMap!![LanguageCode.KOK_IN] = Locale("kok", "IN")

        /// Korean
        languageCodeMap!![LanguageCode.KO_KR] = Locale("ko", "KR")

        /// Kyrgyz - Cyrillic
        languageCodeMap!![LanguageCode.KY_CYRL_KG] = Locale("ky", "CYRL_KG")

        /// Latvian
        languageCodeMap!![LanguageCode.LV_LV] = Locale("lv", "LV")

        /// Lithuanian
        languageCodeMap!![LanguageCode.LT_LT] = Locale("lt", "LT")

        /// Luxembourgish
        languageCodeMap!![LanguageCode.LB_LU] = Locale("lb", "LU")

        /// Macedonian
        languageCodeMap!![LanguageCode.MK_MK] = Locale("mk", "MK")

        /// Malay (Bahasa)
        languageCodeMap!![LanguageCode.MS_MY] = Locale("ms", "MY")

        /// Malayalam (India)
        languageCodeMap!![LanguageCode.ML_IN] = Locale("ml", "IN")

        /// Maltese  (Malta)
        languageCodeMap!![LanguageCode.MT_MT] = Locale("mt", "MT")

        /// Maori - Latin (New Zealand)
        languageCodeMap!![LanguageCode.MI_LATN_NZ] = Locale("mi", "LATN_NZ")

        /// Marathi (India)
        languageCodeMap!![LanguageCode.MR_IN] = Locale("mr", "IN")

        /// Mongolian - Cyrillic
        languageCodeMap!![LanguageCode.MN_CYRL_MN] = Locale("mn", "CYRL_MN")

        /// Nepali (Nepal)
        languageCodeMap!![LanguageCode.NE_NP] = Locale("ne", "NP")

        /// Norwegian (Bokmål)
        languageCodeMap!![LanguageCode.NB_NO] = Locale("nb", "NO")

        /// Norwegian (Nynorsk)
        languageCodeMap!![LanguageCode.NN_NO] = Locale("nn", "NO")

        /// Odia (India)
        languageCodeMap!![LanguageCode.OR_IN] = Locale("or", "IN")

        /// Polish
        languageCodeMap!![LanguageCode.PL_PL] = Locale("pl", "PL")

        /// Portuguese (Brazil)
        languageCodeMap!![LanguageCode.PT_BR] = Locale("pt", "BR")

        /// Portuguese (Portugal)
        languageCodeMap!![LanguageCode.PT_PT] = Locale("pt", "PT")

        /// Punjabi - Gurmukhi
        languageCodeMap!![LanguageCode.PA_GURU] = Locale("pa", "GURU")

        /// Punjabi - Arabic
        languageCodeMap!![LanguageCode.PA_ARAB] = Locale("pa", "ARAB")

        /// Quechua - Latin (Peru)
        languageCodeMap!![LanguageCode.QU_LATN_PE] = Locale("qu", "LATN_PE")

        /// Romanian
        languageCodeMap!![LanguageCode.RO_RO] = Locale("ro", "RO")

        /// Russian
        languageCodeMap!![LanguageCode.RU_RU] = Locale("ru", "RU")

        /// Scottish Gaelic - Latin
        languageCodeMap!![LanguageCode.GD_LATN_GB] = Locale("gd", "LATN_GB")

        /// Serbian - Cyrillic (Bosnia)
        languageCodeMap!![LanguageCode.SR_CYRL_BA] = Locale("sr", "CYRL_BA")

        /// Serbian - Cyrillic (Serbia)
        languageCodeMap!![LanguageCode.SR_CYRL_RS] = Locale("sr", "CYRL_RS")

        /// Serbian - Latin (Serbia)
        languageCodeMap!![LanguageCode.SR_LATN_RS] = Locale("sr", "LATN_RS")

        /// Sesotho Sa Leboa (South Africa)
        languageCodeMap!![LanguageCode.NSO_ZA] = Locale("nso", "ZA")

        /// Setswana
        languageCodeMap!![LanguageCode.TN] = Locale("tn")

        /// Sindhi - Arabic
        languageCodeMap!![LanguageCode.SD_ARAB] = Locale("sd", "ARAB")

        /// Sinhala (Sri Lanka)
        languageCodeMap!![LanguageCode.SI_LK] = Locale("si", "LK")

        /// Slovak
        languageCodeMap!![LanguageCode.SK_SK] = Locale("sk", "SK")

        /// Slovenian
        languageCodeMap!![LanguageCode.SL_SI] = Locale("sl", "SI")

        /// Spanish (Mexico)
        languageCodeMap!![LanguageCode.ES_MX] = Locale("es", "MX")

        /// Spanish (Spain)
        languageCodeMap!![LanguageCode.ES_ES] = Locale("es", "ES")

        /// Swedish
        languageCodeMap!![LanguageCode.SV_SE] = Locale("sv", "SE")

        /// Tajik - Cyrillic
        languageCodeMap!![LanguageCode.TG_CYRL_TJ] = Locale("tg", "CYRL_TJ")

        /// Tamil
        languageCodeMap!![LanguageCode.TA] = Locale("ta")

        /// Tatar - Cyrillic (Russia)
        languageCodeMap!![LanguageCode.TT_CYRL_RU] = Locale("tt", "CYRL_RU")

        /// Telugu (India)
        languageCodeMap!![LanguageCode.TE_IN] = Locale("te", "IN")

        /// Thai
        languageCodeMap!![LanguageCode.TH_TH] = Locale("th", "TH")

        /// Tigrinya (Ethiopia)
        languageCodeMap!![LanguageCode.TI_ET] = Locale("ti", "ET")

        /// Turkish
        languageCodeMap!![LanguageCode.TR_TR] = Locale("tr", "TR")

        /// Turkmen - Latin
        languageCodeMap!![LanguageCode.TK_LATN_TM] = Locale("tk", "LATN_TM")

        /// Ukrainian
        languageCodeMap!![LanguageCode.UK_UA] = Locale("uk", "UA")

        /// Urdu
        languageCodeMap!![LanguageCode.UR] = Locale("ur")

        /// Uyghur - Arabic
        languageCodeMap!![LanguageCode.UG_ARAB] = Locale("ug", "ARAB")

        /// Uzbek - Cyrillic
        languageCodeMap!![LanguageCode.UZ_CYRL_UZ] = Locale("uz", "CYRL_UZ")

        /// Uzbek - Latin
        languageCodeMap!![LanguageCode.UZ_LATN_UZ] = Locale("uz", "LATN_UZ")

        /// Valencian (Spain)
        languageCodeMap!![LanguageCode.CAT_ES] = Locale("cat", "ES")

        /// Vietnamese
        languageCodeMap!![LanguageCode.VI_VN] = Locale("vi", "VN")

        /// Welsh
        languageCodeMap!![LanguageCode.CY_GB] = Locale("cy", "GB")

        /// Wolof - Latin
        languageCodeMap!![LanguageCode.WO_LATN] = Locale("wo", "LATN")

        /// Yoruba - Latin
        languageCodeMap!![LanguageCode.YO_LATN] = Locale("yo", "LATN")
    }
}
